numpy/core/src/multiarray/descriptor.c

```diff
@@ -509,8 +509,11 @@ _convert_from_array_descr(PyObject *obj, int align)
                  && (PyUString_Check(title) || PyUnicode_Check(title))
 #endif
                  && (PyDict_GetItem(fields, title) != NULL))) {
-            PyErr_SetString(PyExc_ValueError,
-                    "two fields with the same name");
+#if defined(NPY_PY3K)
+            name = PyUnicode_AsUTF8String(name);
```

Comment:
`PyUnicode_AsUTF8String` returns a new reference, see [here](https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_AsUTF8String), so you will need to `Py_DECREF(next)` afterwards. There may be other ways, but a possible solution could look like:

```
#if defined(NPY_PY3K)
            name = PyUnicode_AsUTF8String(name);
#else
            Py_INCREF(next);
#endif
            PyErr_Format(PyExc_ValueError,
                    "field '%s' occurs more than once", PyString_AsString(name));
            Py_DECREF(next);
            goto fail;
```

Also, we are assuming that `PyUnicode_AsUTF8String` will always succeed. This seems to be consistent with other uses of the same function in the NumPy code base, but I am not 100% sure it is a good practice.